From f1ee255bc14909990ca437f9e7b0774dd86b299a Mon Sep 17 00:00:00 2001
From: Tobias Schramm <t.schramm@manjaro.org>
Date: Thu, 13 May 2021 16:17:49 +0200
Subject: [PATCH 12/16] spi: spi-bitbang: update spi-mode specific callbacks on
 every transfer

The spi mode used by a device may during initial setup. Previously the
devices spi mode was only evaluated once, resulting in the wrong SPI mode
being used when it changed in the spi slaves probe function.
Fix that by re-evaluating the spi mode during every transfer setup.

Signed-off-by: Tobias Schramm <t.schramm@manjaro.org>
---
 drivers/spi/spi-bitbang.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/spi/spi-bitbang.c b/drivers/spi/spi-bitbang.c
index 1a7352abd878..40c63b31948e 100644
--- a/drivers/spi/spi-bitbang.c
+++ b/drivers/spi/spi-bitbang.c
@@ -140,6 +140,7 @@ int spi_bitbang_setup_transfer(struct spi_device *spi, struct spi_transfer *t)
 	struct spi_bitbang_cs	*cs = spi->controller_state;
 	u8			bits_per_word;
 	u32			hz;
+	struct spi_bitbang	*bitbang = spi_master_get_devdata(spi->master);
 
 	if (t) {
 		bits_per_word = t->bits_per_word;
@@ -149,6 +150,10 @@ int spi_bitbang_setup_transfer(struct spi_device *spi, struct spi_transfer *t)
 		hz = 0;
 	}
 
+	cs->txrx_word = bitbang->txrx_word[spi->mode & (SPI_CPOL|SPI_CPHA)];
+	if (!cs->txrx_word)
+		return -EINVAL;
+
 	/* spi_transfer level calls that work per-word */
 	if (!bits_per_word)
 		bits_per_word = spi->bits_per_word;
-- 
2.30.1

